src/ForeignKeyEditorDelegate.cpp

```diff
@@ -0,0 +1,162 @@
+};
+
+ForeignKeyEditorDelegate::ForeignKeyEditorDelegate(const DBBrowserDB& db, sqlb::Table& table, QObject* parent)
+    : QStyledItemDelegate(parent)
+    , m_db(db)
+    , m_table(table)
+{
+
+}
+
+QWidget* ForeignKeyEditorDelegate::createEditor(QWidget* parent, const QStyleOptionViewItem& option, const QModelIndex& index) const
+{
+    Q_UNUSED(option)
+    Q_UNUSED(index)
+
+    ForeignKeyEditor* editor = new ForeignKeyEditor(parent);
+    editor->setAutoFillBackground(true);
+
+    const auto objects = m_db.getBrowsableObjects();
+    for (auto obj : objects) {
+        if ("table" == obj.gettype()) {
+            QString tableName = obj.table.name();
+            if (tableName == m_table.name())
+                continue;
+
+            m_tablesIds.insert(tableName, obj.table.fieldNames());
+        }
+    }
+
+    editor->tablesComboBox->addItems(m_tablesIds.keys());
+
+    connect(editor->tablesComboBox, static_cast<void(QComboBox::*)(const QString &)>(&QComboBox::currentIndexChanged),
+            [=](const QString& tableName)
+    {
+        QComboBox* box = editor->idsComboBox;
+        box->clear();
+        box->addItems(m_tablesIds[tableName]);
+        box->setCurrentIndex(-1);
+    });
+
+    return editor;
+}
+
+void ForeignKeyEditorDelegate::setEditorData(QWidget* editor, const QModelIndex& index) const
+{
+    ForeignKeyEditor* fkEditor = static_cast<ForeignKeyEditor*>(editor);
+
+    int column = index.row(); // weird? I know right
+    sqlb::FieldPtr field = m_table.fields().at(column);
+    QSharedPointer<sqlb::ForeignKeyClause> fk = m_table.constraint({field}, sqlb::Constraint::ForeignKeyConstraintType).dynamicCast<sqlb::ForeignKeyClause>();
+    if (!fk.isNull()) {
+        fkEditor->tablesComboBox->setCurrentText(fk->table());
+        fkEditor->idsComboBox->setCurrentText(fk->columns().first());
+        fkEditor->clauseEdit->setText(fk->constraint());
+    } else {
+        fkEditor->tablesComboBox->setCurrentIndex(-1);
+    }
+}
+
+void ForeignKeyEditorDelegate::setModelData(QWidget *editor, QAbstractItemModel *model, const QModelIndex &index) const
+{
+    ForeignKeyEditor* fkEditor = static_cast<ForeignKeyEditor*>(editor);
+    QString sql = fkEditor->getSql();
+
+    int column = index.row();
+    sqlb::FieldPtr field = m_table.fields().at(column);
+    if (sql.isEmpty()) {
+        // Remove the foreign key
+        m_table.removeConstraints({field}, sqlb::Constraint::ConstraintTypes::ForeignKeyConstraintType);
+    } else {
+        // Set the foreign key
+        sqlb::ForeignKeyClause* fk = new sqlb::ForeignKeyClause;
+        fk->setTable(fkEditor->tablesComboBox->currentText());
+        fk->setColumns({fkEditor->idsComboBox->currentText()});
+        QString clause = fkEditor->clauseEdit->text();
+        if (!clause.trimmed().isEmpty()) {
+            fk->setConstraint(clause);
+        }
+
+        m_table.setConstraint({field}, sqlb::ConstraintPtr(fk));
+    }
+
+    model->setData(index, sql);
```

Comment:
Maybe need to check for the return value here: if the newly set foreign key won't work, you'll see an error message but afterwards the new foreign key will still be set. This makes it look like it actually did work even though it hasn't. Not sure however what to best do here if false is returned.